client/src/main/java/io/pravega/client/tables/KeyValueTable.java

```diff
@@ -0,0 +1,247 @@
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Defines all operations that are supported on a Key-Value Table.
+ * <p>
+ * A Key-Value Table is a distributed Key-Value Store that indexes Entries by Keys. It uses Table Segments (non-distributed
+ * Key-Value Store backed by a single Pravega Segment) as the fundamental storage primitive and provides a unified view
+ * of all Table Segments involved. Each {@link TableKey} is hashed to a Table Partition which may be represented by one
+ * or more Table Segments (depending on the Key-Value Table configuration chosen when it was created). Such partitioning
+ * enables the Key-Value Table to be distributed across the Pravega cluster but also introduces some constraints for
+ * certain operations (such as multi-key/entry atomic updates). See below for details.
+ * <p>
+ * Key Families are used to group related Keys together in the same Table Partition, which allows multiple
+ * keys/entries belonging to the same Key Family to be updated/removed atomically.
+ * <ul>
+ * <li> Multiple Keys/Entries in the same Key Family can be updated or removed atomically (either all at once or none).
+ * <li> Iterating through all Keys/Entries in the same Key Family is possible.
+ * <li> The same Key may exist in multiple Key Families or even not be associated with any Key Family at all. Such keys
+ * are treated as distinct keys and will not interfere with each other (i.e., if key K1 exists in Key Families F1 and F2,
+ * then F1.K1 is different from F2.K1 and both are different from K1 (no Key Family association).
+ * <li> Keys that do not belong to any Key Family will be uniformly distributed across the Key-Value Table Partitions and
+ * cannot be used for multi-key/entry atomic updates or removals or be iterated on.
+ * <li> Improper use of Key Families may result in degraded performance. If a disproportionate number of Keys/Entries
+ * are placed in the same Key Family (compared to the number of Key/Entries in other Key Families), it may not be
+ * possible to uniformly distribute the Key-Value Table Entries across the cluster and more load will be placed on a
+ * single backing Table Segment instead of spreading such load across many Table Segments. An ideally balanced Key-Value
+ * Table will be one where Keys are not part of any Key Families or the number of Keys in each Key Family is approximately
+ * the same. An improperly designed Key-Value Table will have all Keys part of a single Key Family which will cause a
+ * single Table Segment to bear the full storage and processing load of the entire Key-Value Table.
+ * </ul>
+ * <p>
+ * Types of Updates:
+ * <ul>
+ * <li> Unconditional Updates will insert and/or overwrite any existing values for the given Key, regardless of whether
+ * that Key previously existed or not, and regardless of what that Key's version is.
+ * <li> Conditional Updates will only overwrite an existing value if the specified version matches that Key's version.
```

Comment:
With these conditional and unconditional updates, I think we should blatantly copy the verbage from concurrenthashmap, because they are standardized in the Java Collections.
`put`, `putAll`, `putIfAbsent`, `remove`, and `replace`. 
Java developers will recognise the meaning of these terms.

Reply:
The HashMap/ConcurrentHashMap methods do not have the rich semantics the ones in the KeyValueTableAre. 
`HashMap.put` performs an unconditional update of a single key and returns the existing key.
`HashMap.putAll` performs an unconditional update of multiple key and doesn't return anything.
Similarly with `remove`.

The `KeyValueTable` methods give you more flexibility than just simple `put/putAll/remove`. For example, when you update multiple entries at once, you can set a condition on 0, one or more of those entries; all the conditions on the update batch will need to be satisified in order for the entire batch to be accepted. In addition, you can mix conditions (one key may have no condition, one key may condition on it not existing before, and another key may be conditioned on a specific version). 

HashMap does not give you that - as a programmer, you'd have to write your own code to do something similar.

So I don't see how we can differentiate this using different API names given all the possibilities it supports.

Reply:
Yes, put it unconditional. `replace` is the name of the API when it is conditional on a particular value, and `putIfAbsent` is when it is conditional on the non-existence of the value. I see no reason to name these `put` which is already established as meaning a non-conditional operation.

Reply:
Ok, so how are we going to name the batch operations, which can take multiple keys, each having a condition of a different type (version, not exists, or no condition at all)?

Shall we have `put(single entry)`, `replace(single entry)`, `putIfAbsent(single entry)`, `remove(single key)` (this one for both conditional and unconditional), and `putAll(multiple entries)` and `removeAll`?

Each of these calls will need to validate their input properly. Currently all APIs take in `TableKey` or `TableEntry` which contain various fields, including a version. We'll need to ensure that we don't pass one of these with no version to a conditional API or viceversa.

Reply:
No. I am proposing the following:
Rename KeyVersion to Version and move it off of key and put in on TableEntry. (It has to do with entries no keys after all) and have TableEntry extend Map.Entry.
Then on this class have:
`CompletableFuture<TableEntry<KeyT, ValueT>> get(String keyFamily, KeyT key);`
`CompletableFuture<List<TableEntry<KeyT, ValueT>>> getAll(String keyFamily, List<KeyT> keys);`
`CompletableFuture<Version> put(String keyFamily, KeyT key, ValueT value);`
`CompletableFuture<List<Version>> putAll(String keyFamily, Iterable<Entry<KeyT, ValueT> items);`
`CompletableFuture<Version> putIfAbsent(String keyFamily, KeyT key, ValueT value);`
`CompletableFuture<Version> replace(String keyFamily, KeyT key, Version version, ValueT value);`
`CompletableFuture<List<Version>> replaceAll(String keyFamily, Iterable<TableEntry<KeyT, ValueT>> items);`
`CompletableFuture<TableEntry<KeyT, ValueT>> remove(String keyFamily, KeyT key);`
`CompletableFuture<Boolean> remove(String keyFamily, KeyT key, Version version);`
`CompletableFuture<Boolean> removeAll(String keyFamily, Iterable<TableEntry<KeyT, ValueT>> items);`
In all of the above `keyFamily` may be null.

Reply:
Fixed.